#!/bin/bash

# Load modules
module load tools
module load ngs tools
module load java/1.8.0
module load jre/1.8.0
module load perl
module load gatk/3.8-0
module load samtools/1.7
module load sambamba/0.6.7
module load picard-tools/2.18.0

# Define directories
REF_DIR="/home/projects/pr_46630/data/references/gdc_grch38"
REF_GENOME="$REF_DIR/GRCh38.d1.vd1.fa"
INPUT_DIR="/home/projects/pr_46630/data/Olivera_work/new_samples_thesis/alignment/sorted_bam_files"
OUTPUT_DIR="/home/projects/pr_46630/data/Olivera_work/new_samples_thesis/alignment"
DEDUP_DIR="$OUTPUT_DIR/deduplicated_bam"
REALIGNED_DIR="$OUTPUT_DIR/realigned_bam"
RECALIBRATED_DIR="$OUTPUT_DIR/recalibrated_bam"
KNOWN_VARS_DIR="/home/projects/pr_46630/data/references/variantCalling_references/grch38"

mkdir -p "$DEDUP_DIR" "$REALIGNED_DIR" "$RECALIBRATED_DIR"

#  Marking Duplicates
for BAM in $INPUT_DIR/*.sorted.bam; do
    SAMPLE=$(basename "$BAM" .sorted.bam)
    java -jar /services/tools/picard-tools/2.18.0/picard.jar MarkDuplicates I="$BAM" O="$DEDUP_DIR/${SAMPLE}_dedup.bam" M="$DEDUP_DIR/${SAMPLE}_metrics.txt" REMOVE_DUPLICATES=true
done

for BAM in "$DEDUP_DIR"/*_dedup.bam; do
    SAMPLE=$(basename "$BAM" _dedup.bam)

# Add Read Groups
    java -jar /services/tools/picard-tools/2.18.0/picard.jar AddOrReplaceReadGroups \
        I="$BAM" \
        O="$DEDUP_DIR/${SAMPLE}_dedup_RG.bam" \
        RGID="$SAMPLE" \
        RGLB="lib1" \
        RGPL="ILLUMINA" \
        RGPU="unit1" \
        RGSM="$SAMPLE"

# Index the new BAM file
    samtools index "$DEDUP_DIR/${SAMPLE}_dedup_RG.bam"
done

# Local Indel Realignment
for BAM in $DEDUP_DIR/*_dedup_RG.bam; do
    SAMPLE=$(basename "$BAM" _dedup_RG.bam)

# Create target intervals for realignment
    java -jar /services/tools/gatk/3.8-0/GenomeAnalysisTK.jar -T RealignerTargetCreator \
        -R "$REF_GENOME" \
        -I "$BAM" \
        -known "$KNOWN_VARS_DIR/Homo_sapiens_assembly38.dbsnp138.vcf" \
        -known "$KNOWN_VARS_DIR/Homo_sapiens_assembly38.known_indels.vcf.gz" \
        -o "$REALIGNED_DIR/${SAMPLE}_realign_targets.list"

  # Perform the realignment using the targets
    java -jar /services/tools/gatk/3.8-0/GenomeAnalysisTK.jar -T IndelRealigner \
        -R "$REF_GENOME" \
        -I "$BAM" \
        -known "$KNOWN_VARS_DIR/Homo_sapiens_assembly38.dbsnp138.vcf" \
        -known "$KNOWN_VARS_DIR/Homo_sapiens_assembly38.known_indels.vcf.gz" \
        -targetIntervals "$REALIGNED_DIR/${SAMPLE}_realign_targets.list" \
        -o "$REALIGNED_DIR/${SAMPLE}_realigned.bam"
done

# Base Quality Score Recalibration

for BAM in $REALIGNED_DIR/*_realigned.bam; do
    SAMPLE=$(basename "$BAM" _realigned.bam)

# Generate recalibration data
    java -jar /services/tools/gatk/3.8-0/GenomeAnalysisTK.jar -T BaseRecalibrator \
       -R "$REF_GENOME" \
       -I "$BAM" \
       --knownSites "$KNOWN_VARS_DIR/Homo_sapiens_assembly38.dbsnp138.vcf" \
       --knownSites "$KNOWN_VARS_DIR/Homo_sapiens_assembly38.known_indels.vcf.gz" \
       -o "$RECALIBRATED_DIR/${SAMPLE}_recal_data.table" \
       -nct 4

# Apply the recalibration to the reads
    java -jar /services/tools/gatk/3.8-0/GenomeAnalysisTK.jar -T PrintReads \
       -R "$REF_GENOME" \
       -I "$BAM" -BQSR "$RECALIBRATED_DIR/${SAMPLE}_recal_data.table" \
       -o "$RECALIBRATED_DIR/${SAMPLE}_recalibrated.bam" \
       -nct 4 
done

echo "Complete"
